File: packages/health_connector_hk_ios/ios/health_connector_hk_ios/Sources/health_connector_hk_ios/mappers/health_record_mappers/ExerciseTypeMapper.swift
import Foundation
import HealthKit

/// Extension for converting between `HKWorkoutActivityType` and `ExerciseTypeDto`.
///
/// This mapper implements the iOS HealthKit-specific mapping logic for exercise types
/// as defined in the core `ExerciseType` enum and the Dart iOS mapper logic.
///
/// ## Platform-Specific Behavior
///
/// Not all `ExerciseType` values are supported on iOS HealthKit:
/// - **Android-only types** (annotated with `@supportedOnHealthConnect` in core):
///   These types are NOT present in `ExerciseTypeDto` for iOS and should throw
///   `ArgumentError` if attempted to be used from Dart layer.
/// - **iOS-only types** (annotated with `@supportedOnAppleHealth` in core):
///   These types ARE present in `ExerciseTypeDto` and map to specific `HKWorkoutActivityType` values.
/// - **Cross-platform types**: Supported on both platforms with appropriate mappings.
/// Extension for mapping `HKWorkoutActivityType` ↔ `ExerciseTypeDto`.
extension HKWorkoutActivityType {
    /// Converts `HKWorkoutActivityType` to `ExerciseTypeDto`.
    func toDto() -> ExerciseTypeDto {
        switch self {
        // MARK: - Cardio & Walking/Running (Cross-platform)

        case .running:
            return .running
        case .walking:
            return .walking
        case .cycling:
            return .cycling
        case .hiking:
            return .hiking
        case .handCycling:
            return .handCycling
        case .trackAndField:
            return .trackAndField
        case .swimming:
            return .swimming
        case .surfingSports:
            return .surfing
        case .waterPolo:
            return .waterPolo
        case .rowing:
            return .rowing
        case .sailing:
            return .sailing
        case .paddleSports:
            return .paddling

        // MARK: - Strength Training
        // Cross-platform: strengthTraining (multiple HK types map to this)
        case .traditionalStrengthTraining:
            return .strengthTraining
        case .functionalStrengthTraining:
            return .strengthTraining

        // MARK: - Team Sports
        case .basketball:
            return .basketball
        case .soccer:
            return .soccer
        case .americanFootball:
            return .americanFootball
        case .australianFootball:
            return .australianFootball
        case .baseball:
            return .baseball
        case .softball:
            return .softball
        case .volleyball:
            return .volleyball
        case .rugby:
            return .rugby
        case .cricket:
            return .cricket
        case .handball:
            return .handball
        case .hockey:
            return .hockey
        case .lacrosse:
            return .lacrosse
        case .discSports:
            return .discSports

        // MARK: - Racquet Sports
        case .tennis:
            return .tennis
        case .tableTennis:
            return .tableTennis
        case .badminton:
            return .badminton
        case .squash:
            return .squash
        case .racquetball:
            return .racquetball
        case .pickleball:
            return .pickleball

        // MARK: - Winter Sports
        case .downhillSkiing:
            return .downhillSkiing
        case .crossCountrySkiing:
            return .crossCountrySkiing
        case .snowboarding:
            return .snowboarding
        // Cross-platform: skating (HK uses skatingSports)
        case .skatingSports:
            return .skating
        case .curling:
            return .curling
        case .snowSports:
            return .snowSports

        // MARK: - Martial Arts & Combat Sports
        case .boxing:
            return .boxing
        case .kickboxing:
            return .kickboxing
        case .martialArts:
            return .martialArts
        case .wrestling:
            return .wrestling
        case .fencing:
            return .fencing
        case .taiChi:
            return .taiChi

        // MARK: - Dance & Gymnastics
        // Note: .dance is deprecated in HK, map to .cardioDance
        case .dance:
            return .cardioDance
        case .gymnastics:
            return .gymnastics
        case .socialDance:
            return .socialDance
        case .cardioDance:
            return .cardioDance
        case .barre:
            return .barre

        // MARK: - Fitness & Conditioning
        case .yoga:
            return .yoga
        case .pilates:
            return .pilates
        case .highIntensityIntervalTraining:
            return .highIntensityIntervalTraining
        case .elliptical:
            return .elliptical
        // Cross-platform: stairClimbing (HK also has .stairs which maps to same)
        case .stairClimbing:
            return .stairClimbing
        case .stairs:
            return .stairClimbing
        case .flexibility:
            return .flexibility
        case .stepTraining:
            return .stepTraining
        case .fitnessGaming:
            return .fitnessGaming
        case .coreTraining:
            return .coreTraining
        case .cooldown:
            return .cooldown
        case .mixedCardio:
            return .mixedCardio
        case .crossTraining:
            return .crossTraining
        case .jumpRope:
            return .jumpRope
        case .mindAndBody:
            return .mindAndBody
        case .preparationAndRecovery:
            return .preparationAndRecovery

        // MARK: - Golf & Precision Sports
        case .golf:
            return .golf
        case .archery:
            return .archery
        case .bowling:
            return .bowling

        // MARK: - Outdoor & Adventure
        case .climbing:
            return .climbing
        case .equestrianSports:
            return .equestrianSports
        case .fishing:
            return .fishing
        case .hunting:
            return .hunting
        case .play:
            return .play

        // MARK: - Wheelchair Activities
        case .wheelchairRunPace:
            return .wheelchairRunPace
        case .wheelchairWalkPace:
            return .wheelchairWalkPace

        // MARK: - Multisport
        case .transition:
            if #available(iOS 16.0, *) {
                return .transition
            } else {
                return .other
            }
        case .swimBikeRun:
            if #available(iOS 16.0, *) {
                return .swimBikeRun
            } else {
                return .other
            }

        // MARK: - Other
        case .other:
            return .other
        @unknown default:
            return .other
        }
    }

    /// Converts `ExerciseTypeDto` to `HKWorkoutActivityType`.
    ///
    /// - Parameter dto: The exercise type DTO to convert
    /// - Returns: The corresponding HKWorkoutActivityType
    static func from(dto: ExerciseTypeDto) -> HKWorkoutActivityType {
        switch dto {
        // MARK: - Cardio & Walking/Running

        case .running:
            .running
        case .walking:
            .walking
        case .cycling:
            .cycling
        case .hiking:
            .hiking
        case .handCycling:
            .handCycling
        case .trackAndField:
            .trackAndField

        // MARK: - Water Sports
        case .swimming:
            .swimming
        case .surfing:
            .surfingSports
        case .waterPolo:
            .waterPolo
        case .rowing:
            .rowing
        case .sailing:
            .sailing
        case .paddling:
            .paddleSports
        case .diving:
            if #available(iOS 17.0, *) {
                .underwaterDiving
            } else {
                .other
            }
        case .waterFitness:
            .waterFitness
        case .waterSports:
            .waterSports

        // MARK: - Strength Training
        case .strengthTraining:
            .traditionalStrengthTraining // Use traditional as default

        // MARK: - Team Sports
        case .basketball:
            .basketball
        case .soccer:
            .soccer
        case .americanFootball:
            .americanFootball
        case .australianFootball:
            .australianFootball
        case .baseball:
            .baseball
        case .softball:
            .softball
        case .volleyball:
            .volleyball
        case .rugby:
            .rugby
        case .cricket:
            .cricket
        case .handball:
            .handball
        case .hockey:
            .hockey
        case .lacrosse:
            .lacrosse
        case .discSports:
            .discSports
        case .frisbeeDisc:
            .discSports

        // MARK: - Racquet Sports
        case .tennis:
            .tennis
        case .tableTennis:
            .tableTennis
        case .badminton:
            .badminton
        case .squash:
            .squash
        case .racquetball:
            .racquetball
        case .pickleball:
            .pickleball

        // MARK: - Winter Sports
        case .snowboarding:
            .snowboarding
        case .skating:
            .skatingSports
        case .crossCountrySkiing:
            .crossCountrySkiing
        case .curling:
            .curling
        case .downhillSkiing:
            .downhillSkiing
        case .snowSports:
            .snowSports

        // MARK: - Martial Arts & Combat Sports
        case .boxing:
            .boxing
        case .martialArts:
            .martialArts
        case .fencing:
            .fencing
        case .kickboxing:
            .kickboxing
        case .wrestling:
            .wrestling
        case .taiChi:
            .taiChi

        // MARK: - Dance & Gymnastics
        case .gymnastics:
            .gymnastics
        case .barre:
            .barre
        case .cardioDance:
            .cardioDance
        case .socialDance:
            .socialDance

        // MARK: - Fitness & Conditioning
        case .yoga:
            .yoga
        case .pilates:
            .pilates
        case .highIntensityIntervalTraining:
            .highIntensityIntervalTraining
        case .elliptical:
            .elliptical
        case .stairClimbing:
            .stairClimbing
        case .flexibility:
            .flexibility
        case .crossTraining:
            .crossTraining
        case .jumpRope:
            .jumpRope
        case .stepTraining:
            .stepTraining
        case .fitnessGaming:
            .fitnessGaming
        case .coreTraining:
            .coreTraining
        case .cooldown:
            .cooldown
        case .mixedCardio:
            .mixedCardio
        case .mindAndBody:
            .mindAndBody
        case .preparationAndRecovery:
            .preparationAndRecovery
        case .handCycling:
            .handCycling

        // MARK: - Golf & Precision Sports
        case .golf:
            .golf
        case .archery:
            .archery
        case .bowling:
            .bowling

        // MARK: - Outdoor & Adventure
        case .climbing:
            .climbing
        case .equestrianSports:
            .equestrianSports
        case .fishing:
            .fishing
        case .hunting:
            .hunting
        case .play:
            .play

        // MARK: - Wheelchair Activities
        case .wheelchairWalkPace:
            .wheelchairWalkPace
        case .wheelchairRunPace:
            .wheelchairRunPace

        // MARK: - Multisport
        case .transition:
            if #available(iOS 16.0, *) {
                .transition
            } else {
                .other
            }
        case .swimBikeRun:
            if #available(iOS 16.0, *) {
                .swimBikeRun
            } else {
                .other
            }

        // MARK: - Other
        case .other:
            .other
        }
    }
}

File: packages/health_connector_hk_ios/lib/src/mappers/health_record_mappers/exercise/exercise_type_mapper.dart
import 'package:health_connector_core/health_connector_core_internal.dart'
    show ExerciseType;
import 'package:health_connector_hk_ios/src/pigeon/health_connector_hk_ios_api.g.dart'
    show ExerciseTypeDto;
import 'package:meta/meta.dart' show internal;

/// Converts [ExerciseType] to [ExerciseTypeDto].
@internal
extension ExerciseTypeToDto on ExerciseType {
  ExerciseTypeDto toDto() {
    return switch (this) {
      ExerciseType.other => ExerciseTypeDto.other,
      // Cardio & Walking/Running
      ExerciseType.running => ExerciseTypeDto.running,
      ExerciseType.walking => ExerciseTypeDto.walking,
      ExerciseType.cycling => ExerciseTypeDto.cycling,
      ExerciseType.hiking => ExerciseTypeDto.hiking,

      // Water Sports
      ExerciseType.swimming => ExerciseTypeDto.swimming,
      ExerciseType.surfing => ExerciseTypeDto.surfing,
      ExerciseType.waterPolo ||
      ExerciseType.waterFitness ||
      ExerciseType.waterSports =>
        ExerciseTypeDto.waterPolo, // Fallback or known

      ExerciseType.rowing => ExerciseTypeDto.rowing,
      ExerciseType.sailing => ExerciseTypeDto.sailing,
      ExerciseType.paddling => ExerciseTypeDto.paddling,
      ExerciseType.diving => ExerciseTypeDto.diving,
      // Strength Training
      ExerciseType.strengthTraining => ExerciseTypeDto.strengthTraining,

      // Team Sports
      ExerciseType.basketball => ExerciseTypeDto.basketball,
      ExerciseType.soccer => ExerciseTypeDto.soccer,
      ExerciseType.americanFootball ||
      ExerciseType.frisbeeDisc => ExerciseTypeDto.americanFootball,
      ExerciseType.australianFootball => ExerciseTypeDto.australianFootball,
      ExerciseType.baseball => ExerciseTypeDto.baseball,
      ExerciseType.softball => ExerciseTypeDto.softball,
      ExerciseType.volleyball => ExerciseTypeDto.volleyball,
      ExerciseType.rugby => ExerciseTypeDto.rugby,
      ExerciseType.cricket => ExerciseTypeDto.cricket,
      ExerciseType.handball => ExerciseTypeDto.handball,
      ExerciseType.hockey => ExerciseTypeDto.hockey,

      // Racquet Sports
      ExerciseType.tennis => ExerciseTypeDto.tennis,
      ExerciseType.tableTennis => ExerciseTypeDto.tableTennis,
      ExerciseType.badminton => ExerciseTypeDto.badminton,
      ExerciseType.squash => ExerciseTypeDto.squash,
      ExerciseType.racquetball => ExerciseTypeDto.racquetball,
      ExerciseType.pickleball => ExerciseTypeDto.pickleball,

      // Winter Sports
      ExerciseType.snowboarding => ExerciseTypeDto.snowboarding,
      ExerciseType.skating => ExerciseTypeDto.skating,
      ExerciseType.crossCountrySkiing => ExerciseTypeDto.crossCountrySkiing,
      ExerciseType.downhillSkiing => ExerciseTypeDto.downhillSkiing,
      ExerciseType.snowSports => ExerciseTypeDto.snowSports,
      // Martial Arts & Combat Sports
      ExerciseType.boxing || ExerciseType.kickboxing => ExerciseTypeDto.boxing,
      ExerciseType.martialArts => ExerciseTypeDto.martialArts,
      ExerciseType.wrestling => ExerciseTypeDto.wrestling,
      ExerciseType.fencing => ExerciseTypeDto.fencing,
      ExerciseType.taiChi => ExerciseTypeDto.taiChi,
      // Dance & Gymnastics
      ExerciseType.gymnastics => ExerciseTypeDto.gymnastics,
      ExerciseType.barre => ExerciseTypeDto.barre,
      ExerciseType.cardioDance => ExerciseTypeDto.cardioDance,
      ExerciseType.socialDance => ExerciseTypeDto.socialDance,

      // Fitness & Conditioning
      ExerciseType.yoga => ExerciseTypeDto.yoga,
      ExerciseType.pilates => ExerciseTypeDto.pilates,
      ExerciseType.highIntensityIntervalTraining =>
        ExerciseTypeDto.highIntensityIntervalTraining,
      ExerciseType.crossTraining => ExerciseTypeDto.crossTraining,
      ExerciseType.jumpRope => ExerciseTypeDto.jumpRope,
      ExerciseType.elliptical => ExerciseTypeDto.elliptical,
      ExerciseType.stairClimbing => ExerciseTypeDto.stairClimbing,

      ExerciseType.stepTraining => ExerciseTypeDto.stepTraining,
      ExerciseType.fitnessGaming => ExerciseTypeDto.fitnessGaming,
      ExerciseType.coreTraining => ExerciseTypeDto.coreTraining,
      ExerciseType.flexibility => ExerciseTypeDto.flexibility,
      ExerciseType.cooldown => ExerciseTypeDto.cooldown,
      ExerciseType.mixedCardio => ExerciseTypeDto.mixedCardio,
      ExerciseType.handCycling => ExerciseTypeDto.handCycling,
      ExerciseType.mindAndBody => ExerciseTypeDto.mindAndBody,
      ExerciseType.preparationAndRecovery =>
        ExerciseTypeDto.preparationAndRecovery,
      // Golf & Precision Sports
      ExerciseType.golf => ExerciseTypeDto.golf,
      ExerciseType.discSports => ExerciseTypeDto.discSports,
      ExerciseType.archery => ExerciseTypeDto.archery,
      ExerciseType.bowling => ExerciseTypeDto.bowling,
      ExerciseType.curling => ExerciseTypeDto.curling,
      ExerciseType.lacrosse => ExerciseTypeDto.lacrosse,
      ExerciseType.trackAndField => ExerciseTypeDto.trackAndField,

      // Outdoor & Adventure
      ExerciseType.climbing => ExerciseTypeDto.climbing,
      ExerciseType.equestrianSports => ExerciseTypeDto.equestrianSports,
      ExerciseType.fishing => ExerciseTypeDto.fishing,
      ExerciseType.hunting => ExerciseTypeDto.hunting,
      ExerciseType.play => ExerciseTypeDto.play,
      // Wheelchair Activities
      ExerciseType.wheelchairWalkPace => ExerciseTypeDto.wheelchairWalkPace,
      ExerciseType.wheelchairRunPace => ExerciseTypeDto.wheelchairRunPace,

      // Multisport
      ExerciseType.transition => ExerciseTypeDto.transition,
      ExerciseType.swimBikeRun => ExerciseTypeDto.swimBikeRun,

      // Android Health Connect only types
      ExerciseType.runningTreadmill ||
      ExerciseType.cyclingStationary ||
      ExerciseType.swimmingOpenWater ||
      ExerciseType.swimmingPool ||
      ExerciseType.weightlifting ||
      ExerciseType.calisthenics ||
      ExerciseType.iceHockey ||
      ExerciseType.rollerHockey ||
      ExerciseType.skiing ||
      ExerciseType.snowshoeing ||
      ExerciseType.dancing ||
      ExerciseType.exerciseClass ||
      ExerciseType.bootCamp ||
      ExerciseType.guidedBreathing ||
      ExerciseType.paragliding ||
      ExerciseType.wheelchair => throw ArgumentError(
        'Exercise type $this is not supported on iOS HealthKit',
      ),
    };
  }
}

/// Converts [ExerciseTypeDto] to [ExerciseType].
@internal
extension ExerciseTypeDtoToDomain on ExerciseTypeDto {
  ExerciseType toDomain() {
    return switch (this) {
      ExerciseTypeDto.other => ExerciseType.other,
      // Cardio & Walking/Running
      ExerciseTypeDto.running => ExerciseType.running,
      ExerciseTypeDto.walking => ExerciseType.walking,
      ExerciseTypeDto.cycling => ExerciseType.cycling,
      ExerciseTypeDto.hiking => ExerciseType.hiking,
      // Water Sports
      ExerciseTypeDto.swimming => ExerciseType.swimming,
      ExerciseTypeDto.surfing => ExerciseType.surfing,
      ExerciseTypeDto.waterPolo => ExerciseType.waterPolo,
      ExerciseTypeDto.rowing => ExerciseType.rowing,
      ExerciseTypeDto.sailing => ExerciseType.sailing,
      ExerciseTypeDto.paddling => ExerciseType.paddling,
      ExerciseTypeDto.diving => ExerciseType.diving,
      ExerciseTypeDto.waterFitness => ExerciseType.waterFitness,
      ExerciseTypeDto.waterSports => ExerciseType.waterSports,
      // Strength Training
      ExerciseTypeDto.strengthTraining => ExerciseType.strengthTraining,
      // Team Sports
      ExerciseTypeDto.basketball => ExerciseType.basketball,
      ExerciseTypeDto.soccer => ExerciseType.soccer,
      ExerciseTypeDto.americanFootball => ExerciseType.americanFootball,
      ExerciseTypeDto.frisbeeDisc => ExerciseType.frisbeeDisc,
      ExerciseTypeDto.australianFootball => ExerciseType.australianFootball,
      ExerciseTypeDto.baseball => ExerciseType.baseball,
      ExerciseTypeDto.softball => ExerciseType.softball,
      ExerciseTypeDto.volleyball => ExerciseType.volleyball,
      ExerciseTypeDto.rugby => ExerciseType.rugby,
      ExerciseTypeDto.cricket => ExerciseType.cricket,
      ExerciseTypeDto.handball => ExerciseType.handball,
      ExerciseTypeDto.hockey => ExerciseType.hockey,
      ExerciseTypeDto.lacrosse => ExerciseType.lacrosse,
      ExerciseTypeDto.discSports => ExerciseType.discSports,
      // Racquet Sports
      ExerciseTypeDto.tennis => ExerciseType.tennis,
      ExerciseTypeDto.tableTennis => ExerciseType.tableTennis,
      ExerciseTypeDto.badminton => ExerciseType.badminton,
      ExerciseTypeDto.squash => ExerciseType.squash,
      ExerciseTypeDto.racquetball => ExerciseType.racquetball,
      ExerciseTypeDto.pickleball => ExerciseType.pickleball,
      // Winter Sports
      ExerciseTypeDto.snowboarding => ExerciseType.snowboarding,
      ExerciseTypeDto.skating => ExerciseType.skating,
      ExerciseTypeDto.crossCountrySkiing => ExerciseType.crossCountrySkiing,
      ExerciseTypeDto.curling => ExerciseType.curling,
      ExerciseTypeDto.downhillSkiing => ExerciseType.downhillSkiing,
      ExerciseTypeDto.snowSports => ExerciseType.snowSports,
      // Martial Arts & Combat Sports
      ExerciseTypeDto.boxing => ExerciseType.boxing,
      ExerciseTypeDto.kickboxing => ExerciseType.kickboxing,
      ExerciseTypeDto.martialArts => ExerciseType.martialArts,
      ExerciseTypeDto.wrestling => ExerciseType.wrestling,
      ExerciseTypeDto.fencing => ExerciseType.fencing,
      ExerciseTypeDto.taiChi => ExerciseType.taiChi,
      // Dance & Gymnastics
      ExerciseTypeDto.gymnastics => ExerciseType.gymnastics,
      ExerciseTypeDto.barre => ExerciseType.barre,
      ExerciseTypeDto.cardioDance => ExerciseType.cardioDance,
      ExerciseTypeDto.socialDance => ExerciseType.socialDance,
      // Fitness & Conditioning
      ExerciseTypeDto.yoga => ExerciseType.yoga,
      ExerciseTypeDto.pilates => ExerciseType.pilates,
      ExerciseTypeDto.highIntensityIntervalTraining =>
        ExerciseType.highIntensityIntervalTraining,
      ExerciseTypeDto.elliptical => ExerciseType.elliptical,
      ExerciseTypeDto.stairClimbing => ExerciseType.stairClimbing,
      ExerciseTypeDto.crossTraining => ExerciseType.crossTraining,
      ExerciseTypeDto.jumpRope => ExerciseType.jumpRope,
      ExerciseTypeDto.stepTraining => ExerciseType.stepTraining,
      ExerciseTypeDto.fitnessGaming => ExerciseType.fitnessGaming,
      ExerciseTypeDto.coreTraining => ExerciseType.coreTraining,
      ExerciseTypeDto.flexibility => ExerciseType.flexibility,
      ExerciseTypeDto.cooldown => ExerciseType.cooldown,
      ExerciseTypeDto.mixedCardio => ExerciseType.mixedCardio,
      ExerciseTypeDto.handCycling => ExerciseType.handCycling,
      ExerciseTypeDto.trackAndField => ExerciseType.trackAndField,
      ExerciseTypeDto.mindAndBody => ExerciseType.mindAndBody,
      ExerciseTypeDto.preparationAndRecovery =>
        ExerciseType.preparationAndRecovery,
      // Golf & Precision Sports
      ExerciseTypeDto.golf => ExerciseType.golf,
      ExerciseTypeDto.archery => ExerciseType.archery,
      ExerciseTypeDto.bowling => ExerciseType.bowling,
      // Outdoor & Adventure
      ExerciseTypeDto.climbing => ExerciseType.climbing,
      ExerciseTypeDto.equestrianSports => ExerciseType.equestrianSports,
      ExerciseTypeDto.fishing => ExerciseType.fishing,
      ExerciseTypeDto.hunting => ExerciseType.hunting,
      ExerciseTypeDto.play => ExerciseType.play,
      // Wheelchair Activities
      ExerciseTypeDto.wheelchairWalkPace => ExerciseType.wheelchairWalkPace,
      ExerciseTypeDto.wheelchairRunPace => ExerciseType.wheelchairRunPace,
      // Multisport
      ExerciseTypeDto.transition => ExerciseType.transition,
      ExerciseTypeDto.swimBikeRun => ExerciseType.swimBikeRun,
    };
  }
}
